File: ios/Wavepoint/Analytics/FirebaseCrashReporting.swift
import FirebaseCore
import FirebaseCrashlytics
import Foundation

enum FirebaseCrashReporting {
  static func make(from bundle: Bundle = .main) -> any CrashReporting {
    guard bundle.path(forResource: "GoogleService-Info", ofType: "plist") != nil else {
      return NoOpCrashReporting()
    }
    if FirebaseApp.app() == nil {
      FirebaseApp.configure()
    }
    let crashlytics = Crashlytics.crashlytics()
    crashlytics.setCrashlyticsCollectionEnabled(true)

    return CrashReportingClient { category in
      crashlytics.record(
        error: NSError(
          domain: "ai.mapier.swipe.operational",
          code: category.code,
          userInfo: [NSLocalizedDescriptionKey: category.rawValue]
        )
      )
    }
  }
}

private extension AnalyticsErrorCategory {
  var code: Int {
    switch self {
    case .authorization: 1
    case .eligibility: 2
    case .libraryLoad: 3
    case .commit: 4
    case .playback: 5
    case .configuration: 6
    case .unknown: 7
    }
  }
}

File: android/app/src/main/java/ai/mapier/swipe/analytics/FirebaseCrashReporting.kt
package ai.mapier.swipe.analytics

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.crashlytics.FirebaseCrashlytics

object FirebaseCrashReporting {
  fun make(context: Context): CrashReporting {
    if (FirebaseApp.getApps(context).isEmpty()) return NoOpCrashReporting

    val crashlytics = FirebaseCrashlytics.getInstance()
    crashlytics.setCrashlyticsCollectionEnabled(true)
    return CrashReportingClient { category ->
      crashlytics.recordException(
        WavepointOperationalException(category),
      )
    }
  }
}

private class WavepointOperationalException(category: AnalyticsErrorCategory) :
  RuntimeException(category.value)
